source/common/router/config_impl.cc

```diff
@@ -2147,13 +2146,42 @@ PerFilterConfigs::PerFilterConfigs(
     const Protobuf::Map<std::string, ProtobufWkt::Any>& typed_configs,
     const OptionalHttpFilters& optional_http_filters,
     Server::Configuration::ServerFactoryContext& factory_context,
-    ProtobufMessage::ValidationVisitor& validator) {
-  for (const auto& it : typed_configs) {
-    const auto& name = it.first;
-    auto object = createRouteSpecificFilterConfig(name, it.second, optional_http_filters,
-                                                  factory_context, validator);
-    if (object != nullptr) {
-      configs_[name] = std::move(object);
+    ProtobufMessage::ValidationVisitor& validator)
+    : ignore_optional_option_from_hcm_for_route_config_(Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.ignore_optional_option_from_hcm_for_route_config")) {
+
+  static const std::string filter_config_type =
+      envoy::config::route::v3::FilterConfig::default_instance().GetDescriptor()->full_name();
+
+  for (const auto& per_filter_config : typed_configs) {
+    const std::string& name = per_filter_config.first;
+    RouteSpecificFilterConfigConstSharedPtr config;
+
+    // Ignore the optional_http_filters if the ignore_optional_option_from_hcm_for_route_config_
+    // is set to true by the runtime feature.
+    bool is_optional_by_hcm = !ignore_optional_option_from_hcm_for_route_config_ &&
+                              (optional_http_filters.find(name) != optional_http_filters.end());
+
+    if (TypeUtil::typeUrlToDescriptorFullName(per_filter_config.second.type_url()) ==
+        filter_config_type) {
+      envoy::config::route::v3::FilterConfig filter_config;
+      Envoy::Config::Utility::translateOpaqueConfig(per_filter_config.second, validator,
+                                                    filter_config);
+
+      if (!filter_config.has_config()) {
+        continue;
```

Comment:
> My assumption is that adding a FilterConfig can be used to change some non-specific-filter config (like the disabled field), and that's why it is ok for the code to ignore the entire config.

Yeah, we will implement the `disabled` field in another PR, so, here we just ignore the entire config first.

Reply:
I understand that `disabled` will be added in the future, but I think we are discussing what should happen if `FilterConfig.config` is empty.

At the moment if `FilterConfig.config` is empty, the filter will be skipped. In the future the behavior might be different. An example scenario: The HCM contains a filter that is disabled, and some Routes want to enable it.
If the expectation is that the `FilterConfig.config` could be empty, and the `FilterConfig.disabled` flag will enable it, then the current behavior will break previous configs, no?

Reply:
I don't think it's the case that an empty top-level filter config skips the filter.  The filter itself may be a no-op in that case, but I believe it still gets invoked.

Reply:
If the `FilterConfig.config` is empty, it just means the route level per filter config is nullptr and this route level per filter config will be ignored. But the filter will be enabled as normally. The `FilterConfig` is used to provide route level per filter config, it will not make the filter be skipped.

So, I don't think we will break the behavior.

Reply:
Is there any meaningful usecase for FilterConfig without config specified? If not, we should mark the `config` as required.

Reply:
> It seems to me that an empty-config is a desired property, as it will allow (in the future) to set the filter as disabled for a specific route for example, w/o resending some filter config.

emm....if the config is empty, then there is no way to know which filter you are disabling.

Reply:
> I assume that NACKing was the behavior in Envoy prior to that PR (?), and it's definitely the behavior in gRPC. I don't think we should change that.

It's ok for me to NACK the empty config. Then we need change the `disabled` to a wrappered bool. Then we can NACK the config only when the empty `config` is provided and `disabled` flag is not set explicitly.
It's a break API change, but both envoy and grpc not use it for now, so it should be ok?

WDYT, @markdroth ? cc @adisuissa

Reply:
> It's ok for me to NACK the empty config. Then we need change the `disabled` to a wrappered bool. Then we can NACK the config only when the empty `config` is provided and `disabled` flag is not set explicitly. It's a break API change, but both envoy and grpc not use the `disabled` field for now, so it should be ok?

The change you describe does not break backwards compatibility, as the client will now accept a config which it previously rejected (this is similar to a field that has some valid range, and a PR increases the range - any config sent with the small range will also work with the wider range). Going the other way (decreasing the range) is not allowed.

Reply:
> > I assume that NACKing was the behavior in Envoy prior to that PR (?), and it's definitely the behavior in gRPC. I don't think we should change that.
> 
> I think envoy has no implementation for `FilterConfig` before this PR? My initial idea is that empty `config` + `disabled` could be used to express a filter is diabled explicitly or not.

Right, but the equivalent case can still happen without `FilterConfig`: the value in the `typed_per_filter_config` map can be empty.  I assume Envoy would NACK in that case?

> It's ok for me to NACK the empty config.

Okay, then let's do that.

> Then we need change the `disabled` to a wrappered bool. Then we can NACK the config only when the empty `config` is provided and `disabled` flag is not set explicitly. It's a break API change, but both envoy and grpc not use the `disabled` field for now, so it should be ok?

I don't quite understand what you mean by "wrapped bool".  Can you write exactly what the proto definition would look like?

I'm also not sure why the `config` field needs to be empty in the first place.  Presumably, it could still contain a config of the appropriate type, but it would simply not be used if `disabled` is set.  So control planes could continue to set it, just to make sure they don't break older clients that don't support the `disabled` field.

In any case, I don't think any of this remaining discussion actually needs to block this PR.  Let's move further discussion of this to #27210.

Reply:
I don't think this issue is resolved.  The code still needs to be changed to NACK if the `config` field is unset.

Other than this, though, I have no other concerns with this PR.

Reply:
I will update this PR this weekend.

Reply:
I think this is be resolved now.